Revamping DataDump Tool and adding more operations

nsivabalan opened the pull request:
Revamped DataDump tool and new features have been added

New operations supported:
Compare index entries to Log
Compare replica indexes entries to Log
Compare Log entries to indexes

./gradlew build && ./gradlew test succeeded

SLA: 20 mins
Reviewers: Gopal, Ming

nsivabalan commented on ambry-network/src/test/java/com.github.ambry.network/SSLBlockingChannelTest.java:
```diff
@@ -116,7 +117,7 @@ public void testWrongPortConnection()
       fail("should have thrown!");
     } catch (IOException e) {
-      assertEquals(e.getMessage(), "Connection refused");
+      assertTrue(e.getMessage().contains("Connection refused"));
```
In mac, I got an error as follows:
com.github.ambry.network.SSLBlockingChannelTest > testWrongPortConnection FAILED
    org.junit.ComparisonFailure: expected:<Connection refused[ (Connection refused)]> but was:<Connection refused[]>
        at org.junit.Assert.assertEquals(Assert.java:123)
        at org.junit.Assert.assertEquals(Assert.java:145)
        at com.github.ambry.network.SSLBlockingChannelTest.testWrongPortConnection(SSLBlockingChannelTest.java:119)

hence made this change

vgkholla replied:
It's probably a bad idea to be checking for exception messages in the first place.

nsivabalan replied:
I don't think I wrote this code. So, don't want to change it

vgkholla commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -55,103 +43,134 @@
+ * 6. Dump Log
+ * 7. Dump Replica Metadata token
+ * 8. Compare Index entries to Log entries
+ * 9. Compare Replica Indices entries to Log entries
```
what does this mean?

vgkholla commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,589 @@
+      if (!avoidMiscLogging) {
+        logOutput("version " + version);
+      }
+      if (version == 0) {
```
the only problem I have with this is that changing versions of the index segment (or any other index code) requires changes in tools too.

Once you move these files over to the store package, I recommend reusing `IndexSegment`. The `getEntriesSince()` function might help you achieve what you want without having to rewrite all of that code.

nsivabalan replied:
Yep, will move it in next patch.

vgkholla commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,589 @@
+        randomAccessFile.seek(++tempCurrentOffset);
+        currentOffset = tempCurrentOffset;
+      } catch (EOFException e) {
+        e.printStackTrace();
```
if the exception is going to be thrown, is it necessary to print the stack trace?

nsivabalan commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,589 @@
+      try {
+        short version = randomAccessFile.readShort();
+        if (version == 1) {
+          ByteBuffer buffer = ByteBuffer.allocate(MessageFormatRecord.MessageHeader_Format_V1.getHeaderSize());
```
I do agree that core part where in reading from a log is same, but what we do with it differs in both. I will try to see if I can re-use them

nsivabalan commented:
@vgkholla: Have addressed all your feedbacks. 
WRT testing against large log files, yes, we might potentially see OOM errors. So, we might have to do it in chunks with the help of start and end offset.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -464,384 +451,335 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
+   * @param avoidTraceLogging {@code true} if miscellaneous logging need to be avoided, {@code false} otherwise
+   * @throws Exception
+   */
+  private void compareIndexEntryToLogContent(File indexFile, File replicaDirectory, RandomAccessFile randomAccessFile,
```
Is it necessary to have both this method and `compareIndexEntriestoLogContent`. The two names are also a bit confusing.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -464,384 +451,335 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
+              blobValue.getSize() + " Original Message Offset " + blobValue.getOriginalMessageOffset() +
+              " Flag " + blobValue.getFlags() + "\n";
+          boolean isDeleted = blobValue.isFlagSet(IndexValue.Flags.Delete_Index);
+          if (!isDeleted) {
```
Why only check if the entry is not deleted?

nsivabalan replied:
Good catch. will fix it

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -464,384 +451,335 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
-              messageheader + "\n " + parsedBlobId + "\n" + blobProperty + "\n" + usermetadata + "\n" + blobOutput);
-        } else {
-          System.out.println(messageheader + "\n " + parsedBlobId + "\n" + deleteMsg);
+        if (!avoidTraceLogging) {
```
How about checking `avoidTraceLogging` at the logging method level?

nsivabalan replied:
Nope, it depends on where what each operation wants to log

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -464,384 +451,335 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
-      if (outFile == null) {
-        System.out.println(msg);
+  /**
+   * Compares every entry in the log to those in the index. Checks to see if the status of the blob is consistency with
```
`consistency` -> `consistent`

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -464,384 +451,335 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
+      boolean avoidTraceLogging, boolean filter, boolean generateBlobStatusReport)
+      throws Exception {
+    if (logFileToDump == null || replicaRootDirectory == null) {
+      System.out.println("logFileToDump and replicaRootDirectory needs to be set for compareLogToIndex");
```
For curiosity, when `println` is used, and when `log` is used?

nsivabalan commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -55,103 +43,134 @@
+ * Supported operations are
+ * 1. Dump Index
+ * 2. Dump Index for a replica
+ * 3. Dump active blobs for an index
```
Active are those which are not deleted nor expired.

nsivabalan commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -55,103 +43,134 @@
 
       ArgumentAcceptingOptionSpec<String> outFileOpt =
-          parser.accepts("outFile", "Output file to redirect the output ").withRequiredArg().describedAs("outFile")
+          parser.accepts("outFile", "Output file to redirect the output to")
```
I am not sure why we have this inconsistency. I am also using the same formatter

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -749,6 +768,43 @@ public void compareLogEntriestoIndex(String logFile, ArrayList<String> blobList,
     log("Total Inconsistent blobs count " + totalInconsistentBlobs);
   }
 
+  class MergedIntervals {
```
doc

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+   * @param indexValue the {@link IndexValue} that needs to be compared against
+   * @throws IOException
+   */
+  public boolean readFromLog(RandomAccessFile randomAccessFile, long offset, String blobId, IndexValue indexValue,
```
`readFromLog` -> `readFromLogAndCompare`. Also may need to change the doc for this method.

The reason I am making this suggestion is from the caller of this method, it says compare, but there is no comparison in the caller body. Only until I came to this method I realize that the comparison is done here.

Also, maybe `verify` is better than `compare`?

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+  /**
+   * Holds information about a blob record in the log
+   */
+  class BlobRecordInfo {
```
private class?

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+   * Holds information about a blob record in the log
+   */
+  class BlobRecordInfo {
+    String messageheader = null;
```
-> `messageHeader`

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+    String messageheader = null;
+    BlobId blobId = null;
+    String blobProperty = null;
+    String usermetadata = null;
```
-> `userMetadata`

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+    long timeToLiveInSeconds = -1;
+    int totalRecordSize = 0;
+    short version = randomAccessFile.readShort();
+    if (version == 1) {
```
Do we have only version 1? I remember there is a new version introduced.

nsivabalan replied:
this is header version. We have just one header version for now. Check Message_Header_Version_V1 in MessageFormatRecord

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+      if (coveredRangeInLog != null) {
+        coveredRangeInLog.addInterval(new Pair(offset, offset + blobRecordInfo.totalRecordSize));
+      }
+      compareIndexValueToLogEntry(blobId, indexValue, blobRecordInfo);
```
In the comparison, the actual blob content is not compared. So could we just get the `BlobRecordInfo` without doing `MessageFormatRecord.deserializeBlob(streamlog)`? Simply we can pass a parameter to `readSingleRecordFromLog` indicating if we want to deserialize the actual blob content. I guess this will make the entire comparison much faster.

xiahome replied:
Per our offline discussion, scanning the entire blob content to make sure crc matches. This is more like a verification for no corruption rather than comparing the blob content.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -454,394 +428,362 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
-      if (outFile != null) {
-        fileWriter.flush();
-        fileWriter.close();
+  class MergedIntervals {
```
Doc. And can you make more explanation how this class internally works? How do you check if it is fully covered? This class does not provide a method to do so.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -454,394 +428,362 @@ public long dumpActiveBlobsFromIndex(File indexFileToDump, ArrayList<String> blo
+   * Logs uncovered ranges in the log
+   * @param coveredRangesInLog
+   */
+  private void logUnCoveredRanges(MergedIntervals coveredRangesInLog) {
```
I am a bit confused here, so `coveredRangesInLog` is the entries that match what are in index files. Why you iterate covered intervals to log uncovered ranges?

nsivabalan replied:
Yes, but it refers to those ranges in the log that are not covered via entries in the index. So, I have named it so.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+  /**
+   * Holds the information about a single record in the index
+   */
+  class IndexRecord {
```
private

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+    boolean isDeleted;
+    boolean isExpired;
+
+    public IndexRecord(String msg, boolean isDeleted, boolean isExpired) {
```
Remove `public`

nsivabalan replied:
DataDump uses this class. Hence can't make it private.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpDataHelper.java:
```diff
@@ -0,0 +1,630 @@
+  /**
+   * Holds information about a blob in the log
+   */
+  class LogBlobRecord {
```
private

Also there are some public methods you may want to change to package level.

nsivabalan replied:
DataDump uses this class. Hence can't make it private.

xiahome commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/DumpData.java:
```diff
@@ -731,6 +765,10 @@ public int compare(Pair<Long, Long> o1, Pair<Long, Long> o2) {
+    /**
+     * Adds and merges the interval if need be
+     * @param newInterval the new interval that needs to be added
+     */
```
In (maybe unlikely) case of overlapping intervals, is `MergedInterval` class still able to handle them correctly? Say we have two intervals t1-t2, and t3-t4, and t3>t1, t4<t2, t3<t4(of course). What would be `logUnCoveredRanges` show?

Not sure if there would be an easy fix, just modify the comparator method, so that if there are overlapping intervals, fail the test.

nsivabalan replied:
Nope. as of now MergedInterval class doesn't handle this case. We can fix it going forward if we see such cases.

xiahome replied:
I am ok with fixing in a different patch. This issue should not happen I think.

xiahome commented:
LTGM.
